PLAT-10242 Configuration

symphony-youness opened the pull request:
### PLAT-10242
[PLAT-10242](https://perzoinc.atlassian.net/browse/PLAT-10242)

### Description
Added the configuration model following the model implemented in Java BDK2.0: 
[Java BDK model](https://github.com/SymphonyPlatformSolutions/symphony-api-client-java/tree/master/symphony-bdk-core/src/main/java/com/symphony/bdk/core/config/model)

I didn't implement the parameter interpolation in the BdkConfigParser

The BdkConfig class contains:
- [x] agent
- [x] pod
- [x] key_manager
- [x] session_auth
- [x] bot
- [x] ssl
- [ ] datafeed
- [ ] app
- [ ] retry
- [ ] proxy

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [x] Updated the documentation in [docs folder](../docs)

symphony-youness committed:
PLAT-10242 configuration parser - no interpolation
added configuration parser
updated toml to add `pyyaml` dependency

symphony-youness committed:
PLAT-10242 configuration loader
added configuration loader
added BdkConfig class and models:
- agent, pod, key_manager, session_auth, bot, ssl

symphony-youness committed:
PLAT-10242 added type hints

symphony-youness committed:
PLAT-10242 update poetry.lock
Necessary solve dependencies using poetry

symphony-elias commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,49 @@
+            config_content = config_path.read_text()
+            return cls.load_from_content(config_content)
+        else:
+            raise BdkConfigException("Config file has not been found at: %s" % config_path.absolute())
```
Please use f-strings: https://realpython.com/python-string-formatting/

symphony-elias commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,49 @@
+    def load_from_symphony_dir(cls, relative_path: str):
+        """Load BdkConfig from a relative path located in the .symphony directory.
+
+        Note: The .symphony directory is located under your home directory (``System.getProperty("user.home")``).
```
`System.getProperty("user.home")` is java-specific, IMHO you can remove that as `home directory` is clear enough

symphony-elias commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,49 @@
+        """
+        home_path = Path.home()
+        symphony_dir_path = home_path / ".symphony"
+        config_path = (symphony_dir_path / relative_path).resolve()
```
IMHO, you could inline home_path and symphony_dir_path and put directly: `config_path = (Path.home() / ".symphony" / relative_path).resolve()

symphony-elias commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,49 @@
+        symphony_dir_path = home_path / ".symphony"
+        config_path = (symphony_dir_path / relative_path).resolve()
+
+        if config_path.exists():
```
Please reuse `load_from_file`

symphony-elias commented on symphony/bdk/core/config/bdk_config_parser.py:
```diff
@@ -0,0 +1,25 @@
+        except yaml.YAMLError as e:
+            pass
+
+        raise BdkConfigException("Config file is not the right file type")
```
Please be more specific with something like : `Config file format is neither JSON nor YAML"

symphony-youness replied:
I will change it to `Config file has a wrong format.` as it can be a misformatted YAML file. 
There is a test using an invalid yaml file btw.

symphony-youness replied:
We can add that information in the logs. After implementing that.

symphony-elias commented on symphony/bdk/core/config/model/bdk_agent_config.py:
```diff
@@ -0,0 +1,7 @@
+from symphony.bdk.core.config.model.bdk_client_config import BdkClientConfig
```
If I remember well, this class was used to accommodate for the load-balancing usecase. So you can remove it,

symphony-youness replied:
I shouldn't have pushed this in the first place... 😄

symphony-elias commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,49 @@
+        return cls.parse_config(BdkConfigParser.parse(content))
+
+    @staticmethod
+    def parse_config(data_dict: dict):
```
IMHO this belongs to the BdkConfigParser class.

symphony-youness committed:
PLAT-10242 small fixes
deleted BdkAgentConfig class
use f-strings
moved parse_config logic to config_loader

symphony-youness committed:
PLAT-10242 used load_from_file()
used helper function load_from_file() in load_from_symphony_dir()

symphony-elias commented on symphony/bdk/core/config/model/bdk_authentication_config.py:
```diff
@@ -0,0 +1,31 @@
+
+        :return true if the RSA authentication is configured
+        """
+        return (self.bot_rsa_key_config is not None and self.bot_rsa_key_config.is_configured())
```
`self.bot_rsa_key_config is not None` will always be true because of initialization in constructor.

symphony-elias commented on symphony/bdk/core/config/model/bdk_bot_config.py:
```diff
@@ -0,0 +1,21 @@
+
+    def __init__(self, config):
+        if config is not None:
+            self.__username = config.get("username")
```
Please use a single underscore for the username field.

symphony-elias commented on symphony/bdk/core/config/model/bdk_certificate_config.py:
```diff
@@ -0,0 +1,46 @@
+        """
+        return self.path is not None or self.content
+
+    def is_valid(self) -> bool:
```
Do we need two functions for the same check?

symphony-elias commented on symphony/bdk/core/config/model/bdk_client_config.py:
```diff
@@ -0,0 +1,42 @@
+    def __init__(self, parent_config, config):
+        self.parent_config = parent_config
+        if config is not None:
+            self._scheme = config.get("scheme") if "scheme" in config else None
```
https://docs.python.org/3/library/stdtypes.html#dict.get : get("key") on a dict defaults to `None` id key is not present.

symphony-elias commented on symphony/bdk/core/config/model/bdk_client_config.py:
```diff
@@ -0,0 +1,42 @@
+    def proxy(self):
+        return self.__self_or_parent(self._proxy, self.parent_config.proxy)
+
+    def __self_or_parent(self, instance_value, parent_value):
```
Please use one single underscore for private methods. Please check for the difference between single underscore and double underscores: https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name

symphony-elias commented on symphony/bdk/core/config/model/bdk_config.py:
```diff
@@ -0,0 +1,55 @@
+class BdkConfig(BdkServerConfig):
+    """Class containing the Bdk configuration """
+
+    def __init__(self, **kwargs):
```
Please put directly constructor params `scheme`, `host`, `port`, etc. before **kwargs and remove `kwargs.get`
The code below works:
```python
def _private(a, b, c, **kwargs):
    print(f'a={a}, b={b}, c={c}')
    print(kwargs)

def test_unpack():
    d = {'a': "value a", 'b': "value b", 'c': "value c", 'd': 'toto'}
    _private(**d)
```

symphony-youness replied:
After thinking about it, (And please tell me if I overthought this one) but what if the dictionary looks like this: (removed c value)
```python
d = {'a': "value a", 'b': "value b", 'd': 'toto'}
```
This will raise the error: `TypeError: _private() missing 1 required positional argument: 'c'`

<br>

I don't think this is a use case we want. This means if we write our code like this:
```python
class BdkConfig(...)
def __init__(self, host, context, port, agent, pod, ...,  **kwargs):
```

and don't provide the `context` at the root of the config file such as the minimalist one:
```yaml
host: acme.symphony.com                                    
bot: 
    username: bot-username                                  
    privateKey:
      path: /path/to/bot/rsa-private-key.pem
```
We will get this error, `TypeError: __init__() missing 1 required positional argument: 'context'` which is not something we want, right ?

symphony-elias replied:
What about `context = None` in the parameters ?

symphony-elias commented on symphony/bdk/core/config/model/bdk_config.py:
```diff
@@ -0,0 +1,55 @@
+        self.__ssl = BdkSslConfig(kwargs.get("ssl"))
+
+    @property
+    def agent(self): return self.__agent
```
If we don't need any specific logic in getters and setters, IMHO we can use fields without underscores and remove the properties

symphony-elias commented on symphony/bdk/core/config/model/bdk_server_config.py:
```diff
@@ -0,0 +1,78 @@
+    _DEFAULT_HTTPS_PORT: int = 443
+
+    def __init__(self, scheme=None, port=None, host=None, context=None, **kwargs):
+        self._scheme = scheme if scheme is not None else self._DEFAULT_SCHEME
```
You can put _DEFAULT_SCHEME and _DEFAULT_HTTPS_PORT directly as parameter default values.

symphony-youness replied:
I can't do that because we are passing arguments using `dict.get()` that is passing `None` as an argument which overrides the default parameters.
```python
>>> BdkConfig
 config = {}
super().__init__(scheme=config.get("scheme"))

#Which overrides the default value in 
>>> BdkServerConfig 
def __init__(self, scheme=None)
```

symphony-elias commented on symphony/bdk/core/config/model/bdk_ssl_config.py:
```diff
@@ -0,0 +1,10 @@
+
+    def __init__(self, config):
+        if config is not None:
+            self.trust_store = BdkCertificateConfig(**config.get("trustStore")) if "trustStore" in config else BdkCertificateConfig()
```
Please put a trustStore param in the constructor, we don't need the full config dict.

thibauult commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,40 @@
+from symphony.bdk.core.config.model.bdk_config import BdkConfig
+
+
+class BdkConfigLoader:
```
All classes should be documented

symphony-elias commented on tests/core/config/test_bdk_config_loader.py:
```diff
@@ -0,0 +1,83 @@
+        config = BdkConfigLoader.load_from_file(wrong_path)
+
+
+def test_main_test():
```
Please remove this test

symphony-elias commented on tests/core/config/test_bdk_config_loader.py:
```diff
@@ -0,0 +1,83 @@
+    assert config.bot.username == "youbot"
+
+
+def test_load_client_global_config(global_config_path):
```
Please test get_base_path returns the correct output when we have a configuration at root level and when we don't have it. Or as suggested, drop the override logic with no scheme, host, port, context at root level.

symphony-youness committed:
PLAT-10242 Added examples
Added documentation for the configuration
Fixed most problems
Added class documentation
Added ConfigMain example examples/core/config/

symphony-youness committed:
Merge branch '2.0' of github.com:SymphonyPlatformSolutions/symphony-api-client-python into PLAT-10242

# Conflicts:
#	poetry.lock
#	pyproject.toml

symphony-youri commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,56 @@
+class BdkConfigLoader:
+    """ Config loader class
+
+    Provide methods to load the a JSON or YAML configuration
```
"the a"

symphony-youri commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,56 @@
+    """ Config loader class
+
+    Provide methods to load the a JSON or YAML configuration
+    from an absolute path or `USER_HOME_PATH/.symphony``
```
$HOME is common to point to the home folder (or ~/.symphony)

symphony-youri commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,56 @@
+    """
+
+    @classmethod
+    def load_from_file(cls, config_path: str):
```
should we use python type hints in our public api (and every where maybe?)

symphony-youri commented on symphony/bdk/core/config/bdk_config_loader.py:
```diff
@@ -0,0 +1,56 @@
+            raise BdkConfigException(f"Config file has not been found at: {config_path.absolute()}")
+
+    @classmethod
+    def load_from_content(cls, content: str) -> BdkConfig:
```
oh there we use it

symphony-youri commented on symphony/bdk/core/config/model/bdk_server_config.py:
```diff
@@ -0,0 +1,80 @@
+        local_context = self.context
+        if local_context == None:
+            return ""
+        if local_context and local_context[0] != "/":
```
is it safe to access [0] without checks?

symphony-youness replied:
I don't think it will work if the context is an int ... 
I will add a condition to check if local_context is a string.

thibauult commented on docs/configuration.md:
```diff
@@ -0,0 +1,49 @@
+        with open("/absolute/path/to/config.yaml") as config_file:                 #2
+            config_2 = BdkConfigLoader.load_from_content(config_file.read())
+
+        config_3 = BdkConfigLoader.load_from_symphony_dir("/absolute/path/to/config.yaml") #3
```
Param should be a relative location from the `~/.symphony` directory

thibauult commented on examples/core/config/config_main.py:
```diff
@@ -0,0 +1,20 @@
+        with open("/absolute/path/to/config.yaml") as config_file:
+            config_2 = BdkConfigLoader.load_from_content(config_file.read())
+
+        config_3 = BdkConfigLoader.load_from_symphony_dir("/absolute/path/to/config.yaml")
```
Same comment, path is wrong here

thibauult commented on tests/resources/config/config.json:
```diff
@@ -0,0 +1,17 @@
+  },
+  "bot": {
+    "username": "youbot",
+    "privateKeyPath": "/Users/local/conf/agent/privatekey.pem"
```
`bot` > `privateKey` > `path`

thibauult commented on tests/resources/config/config_global.json:
```diff
@@ -0,0 +1,28 @@
+  },
+  "bot": {
+    "username": "tibot",
+    "privateKeyPath": "/Users/local/conf/agent/privatekey.pem"
```
Python BDK should not support `privateKeyPath` (it's something already deprecated in Java), but `privateKey` > `path` instead.

symphony-youness replied:
this was a mistake. it should not be used.

symphony-youness committed:
removed 'private' attributes field unnecessary properties
removed unnecessary test
fixed typos, wrong paths and config files

symphony-youness committed:
correct configuration files format

symphony-elias commented on pyproject.toml:
```diff
@@ -29,7 +30,6 @@ junit_logging = "all"
 [tool.coverage.run]
 branch = true
 source = ["symphony"]
-omit = ["symphony/bdk/gen/*"]
```
Please keep that as we don't want to run coverage checks on the generated code.

symphony-youness committed:
- changes attributes of authentication config to have attributes names 'private_key' and 'certificatate'
- documented the utils
- tested BdkAuthenticationConfig, BdkClientConfig and BdkServerConfig

symphony-youness committed:
added tests
Added tests for get_base_path when a configuration has not root values (scheme, host ...)

thibauult commented on docs/configuration.md:
````diff
@@ -0,0 +1,49 @@
+
+if __name__ == "__main__":
+    ConfigMain.run()
+```
````
Could we make doc examples a bit simpler: 
```python
from symphony.bdk.core.config.bdk_config_loader import BdkConfigLoader
from symphony.bdk.core.config.model.bdk_config import BdkConfig

if __name__ == "__main__":

  config_1 = BdkConfigLoader.load_from_file("/absolute/path/to/config.yaml") #1

  with open("/absolute/path/to/config.yaml") as config_file:                 #2
    config_2 = BdkConfigLoader.load_from_content(config_file.read())

  config_3 = BdkConfigLoader.load_from_symphony_dir("config.yaml") #3

  config_4 = BdkConfig() #4
```

Let's also try to improve imports. For instance, could we only have: 
```python
from symphony.bdk.core.config import BdkConfigLoader, BdkConfig
```
?

symphony-youness replied:
For the imports. directly like this, no we can't. I will check if we can do smth to fix that.<br>
For the doc we can definitly do this as we don't need a main in the example and it works
```python
config_1 = BdkConfigLoader.load_from_file("/absolute/path/to/config.yaml") #1
config_2 = BdkConfigLoader.load_from_symphony_dir("relative/path/to/config.yaml") #2
config_3 = BdkConfig() #3
```

symphony-elias commented on docs/configuration.md:
```diff
@@ -0,0 +1,49 @@
+
+    @staticmethod
+    def run():
+        #
```
Remove the useless hash

symphony-elias commented on symphony/bdk/core/config/model/bdk_certificate_config.py:
```diff
@@ -0,0 +1,22 @@
+
+        :return: true if the RSA configuration is invalid.
+        """
+        return (self.path is not None and self.content != "") and self.password is not None
```
I guess it is: `(self.path is not None *or* self.content != "") and self.password is not None`, i.e. you have either a path to a cert file or the cert content.

symphony-youness replied:
ah I think I misunderstood that one then.
it's probably 
```python
not (self.path is not None and self.content != "") and self.password is not None
```
to enforce only one being used. <br>
In that aspect I suppose, I should do the same for the rsa private key.

symphony-elias commented on symphony/bdk/core/config/model/bdk_config.py:
```diff
@@ -0,0 +1,21 @@
+    :return self: BdkConfig instance
+    """
+    def __init__(self, **config):
+        super().__init__(scheme=config.get("scheme"), host=config.get("host"), port=config.get("port"),
```
Please do the same as BdkCertificateConfig constructor.

symphony-youness replied:
ok

symphony-youness committed:
fixed authentication config

symphony-youness committed:
changed constructor params

symphony-elias commented on symphony/bdk/core/config/model/bdk_certificate_config.py:
```diff
@@ -0,0 +1,22 @@
+        """Check if the certificate configuration is valid.
+        If both certificate path and content, the configuration is invalid.
+
+        :return: true if the RSA configuration is invalid.
```
`if the RSA configuration is invalid` -> should be `valid` instead.

symphony-elias commented on symphony/bdk/core/config/model/bdk_rsa_key_config.py:
```diff
@@ -0,0 +1,21 @@
+        """Check if the RSA configuration is valid.
+        If both of private key path and content, the configuration is invalid.
+
+        :return: true if the RSA configuration is invalid.
```
Should be `true if the RSA configuration is valid`

symphony-youness committed:
fixed typos
